File: macOS Source/TotallyHuman/Database/TrainingStore.swift
// TrainingStore.swift — Robuster, menschenlesbarer Trainingsdaten-Speicher
// Totally Human — KI-Musik-Detektor (native macOS, kein Python)
//
// Warum eine JSON-Datei statt SQLite?
//   Die frühere SQLite-Anbindung konnte auf manchen Systemen still fehlschlagen
//   (Sandbox / Application-Support-Rechte). Dann wurde beim Import zwar hochgezählt,
//   aber nichts persistiert ("148 importiert, aber 0 in der DB"). Eine einfache,
//   atomar geschriebene JSON-Datei ist robust, transparent und für den Nutzer
//   direkt einsehbar. Bei Schreibfehlern greift ein In-Memory-Fallback, damit die
//   App im laufenden Betrieb niemals in den heuristischen Modus zurückfällt.
//
// Speicherort (lokal, bleibt zwischen Starts erhalten -> Modell lernt dauerhaft dazu):
//   ~/Library/Application Support/Totally Human/trainingsdaten.json

import Foundation
import AppKit

/// Ermittelt EINMALIG einen tatsächlich beschreibbaren Basis-Speicherort und legt
/// ihn sofort an. Dadurch existiert der Ordner ab dem ersten Start — auch wenn das
/// bevorzugte „Application Support"-Verzeichnis (z. B. wegen Rechten/Sandbox) nicht
/// beschreibbar ist. Modell UND Trainingsdaten teilen sich diese Basis, damit beim
/// „Ordner öffnen" wirklich der genutzte Ordner erscheint.
enum AppSpeicherort {

    /// Der final genutzte Basis-Ordner (garantiert beschreibbar oder Temp als Notnagel).
    static let basis: URL = ermittleBasis()

    /// Fehlermeldung, falls der bevorzugte Ort nicht beschreibbar war.
    private(set) static var fehler: String?

    private static func ermittleBasis() -> URL {
        let fm = FileManager.default
        var kandidaten: [URL] = []

        if let appSup = try? fm.url(for: .applicationSupportDirectory,
                                    in: .userDomainMask,
                                    appropriateFor: nil,
                                    create: true) {
            kandidaten.append(appSup.appendingPathComponent("Totally Human", isDirectory: true))
        }
        let home = fm.homeDirectoryForCurrentUser
        kandidaten.append(home.appendingPathComponent("Library/Application Support/Totally Human", isDirectory: true))
        kandidaten.append(home.appendingPathComponent("Documents/Totally Human", isDirectory: true))
        kandidaten.append(fm.temporaryDirectory.appendingPathComponent("Totally Human", isDirectory: true))

        for kandidat in kandidaten {
            if istBeschreibbar(kandidat) {
                fehler = nil
                NSLog("[TotallyHuman] Speicherort: \(kandidat.path)")
                return kandidat
            }
        }
        // Absoluter Notnagel — sollte nie erreicht werden.
        let notnagel = fm.temporaryDirectory.appendingPathComponent("Totally Human", isDirectory: true)
        try? fm.createDirectory(at: notnagel, withIntermediateDirectories: true)
        return notnagel
    }

    /// Legt den Ordner an und prüft mit einer echten Schreibprobe, ob er nutzbar ist.
    private static func istBeschreibbar(_ dir: URL) -> Bool {
        let fm = FileManager.default
        do {
            try fm.createDirectory(at: dir, withIntermediateDirectories: true)
            let probe = dir.appendingPathComponent(".schreibtest")
            try Data("ok".utf8).write(to: probe, options: .atomic)
            try? fm.removeItem(at: probe)
            return true
        } catch {
            fehler = error.localizedDescription
            NSLog("[TotallyHuman] Speicherort nicht beschreibbar (\(dir.path)): \(error)")
            return false
        }
    }
}

final class TrainingStore {

    /// Alle bekannten Trainingsproben (Quelle der Wahrheit im Speicher).
    private(set) var proben: [TrainingsProbe] = []

    /// Wird true, sobald das Laden/Speichern auf die Platte funktioniert hat.
    private(set) var persistenzAktiv: Bool = false

    /// Letzte Fehlermeldung (für die Statusanzeige), falls vorhanden.
    var letzterFehler: String? { eigenerFehler ?? AppSpeicherort.fehler }
    private var eigenerFehler: String?

    // MARK: - Pfad

    /// Vollständiger Pfad der Trainingsdaten-Datei.
    let dateiURL: URL

    /// Der genutzte Basis-Ordner (für „Ordner öffnen").
    var basisOrdner: URL { AppSpeicherort.basis }

    /// Für die UI: menschenlesbarer Pfad des Trainingsdaten-Speichers.
    var pfad: String { dateiURL.path }

    init() {
        // Basis ist bereits angelegt & beschreibbar geprüft -> Ordner existiert ab Start.
        self.dateiURL = AppSpeicherort.basis
            .appendingPathComponent("trainingsdaten.json")
        laden()
    }

    // MARK: - Laden / Speichern

    /// Lädt die Trainingsdaten von der Platte (falls vorhanden).
    func laden() {
        let fm = FileManager.default
        guard fm.fileExists(atPath: dateiURL.path) else {
            // Datei existiert noch nicht — ist beim ersten Start normal.
            // Der Ordner wurde bereits in AppSpeicherort angelegt & getestet.
            persistenzAktiv = (AppSpeicherort.fehler == nil)
            return
        }
        do {
            let data = try Data(contentsOf: dateiURL)
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            proben = try decoder.decode([TrainingsProbe].self, from: data)
            persistenzAktiv = true
            eigenerFehler = nil
            NSLog("[TotallyHuman] Trainingsdaten geladen: \(proben.count) Beispiele aus \(dateiURL.path)")
        } catch {
            eigenerFehler = error.localizedDescription
            NSLog("[TotallyHuman] Trainingsdaten laden fehlgeschlagen: \(error)")
        }
    }

    /// Schreibt die Trainingsdaten atomar auf die Platte.
    @discardableResult
    func speichern() -> Bool {
        let fm = FileManager.default
        do {
            try fm.createDirectory(at: dateiURL.deletingLastPathComponent(),
                                   withIntermediateDirectories: true)
            let encoder = JSONEncoder()
            encoder.dateEncodingStrategy = .iso8601
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            let data = try encoder.encode(proben)
            try data.write(to: dateiURL, options: .atomic)
            persistenzAktiv = true
            eigenerFehler = nil
            return true
        } catch {
            // In-Memory-Fallback: Daten bleiben im Arbeitsspeicher erhalten,
            // damit die App weiterläuft; nur die Persistenz ist beeinträchtigt.
            persistenzAktiv = false
            eigenerFehler = error.localizedDescription
            NSLog("[TotallyHuman] Trainingsdaten speichern fehlgeschlagen: \(error)")
            return false
        }
    }

    // MARK: - Mutationen

    /// Fügt eine Probe hinzu. Dedupliziert anhand der Datei-Signatur.
    /// Gibt true zurück, wenn tatsächlich NEU eingefügt wurde.
    @discardableResult
    func hinzufuegen(_ probe: TrainingsProbe) -> Bool {
        if !probe.dateiSignatur.isEmpty,
           proben.contains(where: { $0.dateiSignatur == probe.dateiSignatur }) {
            return false   // Duplikat
        }
        proben.append(probe)
        speichern()
        return true
    }

    /// Fügt mehrere Proben hinzu. Gibt die Anzahl der wirklich neu eingefügten zurück.
    @discardableResult
    func hinzufuegen(_ neue: [TrainingsProbe]) -> Int {
        var eingefuegt = 0
        for p in neue {
            if !p.dateiSignatur.isEmpty,
               proben.contains(where: { $0.dateiSignatur == p.dateiSignatur }) {
                continue
            }
            proben.append(p)
            eingefuegt += 1
        }
        if eingefuegt > 0 { speichern() }
        return eingefuegt
    }

    /// Löscht alle Trainingsdaten.
    func leeren() {
        proben.removeAll()
        speichern()
    }
}

File: Windows Source/TotallyHuman/Database/TrainingStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using TotallyHuman.Models;
using TotallyHuman.Utilities;

namespace TotallyHuman.Database;

/// <summary>
/// Robuster, menschenlesbarer Trainingsdaten-Speicher (JSON). 1:1 Port von
/// TrainingStore.swift. Speicherort: %APPDATA%\TotallyHuman\trainingsdaten.json
/// </summary>
public sealed class TrainingStore
{
    private readonly List<TrainingsProbe> _proben = new();

    /// <summary>Alle bekannten Trainingsproben (Quelle der Wahrheit im Speicher).</summary>
    public IReadOnlyList<TrainingsProbe> Proben => _proben;

    /// <summary>true, sobald Laden/Speichern auf die Platte funktioniert hat.</summary>
    public bool PersistenzAktiv { get; private set; }

    private string? _eigenerFehler;
    public string? LetzterFehler => _eigenerFehler ?? AppSpeicherort.Fehler;

    public string DateiPfad { get; }
    public string BasisOrdner => AppSpeicherort.Basis;
    public string Pfad => DateiPfad;

    private static readonly JsonSerializerOptions Options = new()
    {
        WriteIndented = true,
        Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
    };

    public TrainingStore()
    {
        DateiPfad = Path.Combine(AppSpeicherort.Basis, "trainingsdaten.json");
        Laden();
    }

    public void Laden()
    {
        if (!File.Exists(DateiPfad))
        {
            PersistenzAktiv = AppSpeicherort.Fehler == null;
            return;
        }
        try
        {
            string json = File.ReadAllText(DateiPfad);
            var geladen = JsonSerializer.Deserialize<List<TrainingsProbe>>(json, Options);
            _proben.Clear();
            if (geladen != null) _proben.AddRange(geladen);
            PersistenzAktiv = true;
            _eigenerFehler = null;
        }
        catch (Exception ex)
        {
            _eigenerFehler = ex.Message;
        }
    }

    public bool Speichern()
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(DateiPfad)!);
            string json = JsonSerializer.Serialize(_proben, Options);
            string tmp = DateiPfad + ".tmp";
            File.WriteAllText(tmp, json);
            if (File.Exists(DateiPfad)) File.Delete(DateiPfad);
            File.Move(tmp, DateiPfad);
            PersistenzAktiv = true;
            _eigenerFehler = null;
            return true;
        }
        catch (Exception ex)
        {
            PersistenzAktiv = false;
            _eigenerFehler = ex.Message;
            return false;
        }
    }

    /// <summary>Fügt eine Probe hinzu (dedupliziert via Datei-Signatur). true = neu eingefügt.</summary>
    public bool Hinzufuegen(TrainingsProbe probe)
    {
        if (!string.IsNullOrEmpty(probe.DateiSignatur) &&
            _proben.Any(p => p.DateiSignatur == probe.DateiSignatur))
            return false;
        _proben.Add(probe);
        Speichern();
        return true;
    }

    /// <summary>Fügt mehrere Proben hinzu. Gibt die Anzahl wirklich neu eingefügter zurück.</summary>
    public int Hinzufuegen(IEnumerable<TrainingsProbe> neue)
    {
        int eingefuegt = 0;
        foreach (var p in neue)
        {
            if (!string.IsNullOrEmpty(p.DateiSignatur) &&
                _proben.Any(x => x.DateiSignatur == p.DateiSignatur))
                continue;
            _proben.Add(p);
            eingefuegt++;
        }
        if (eingefuegt > 0) Speichern();
        return eingefuegt;
    }

    public void Leeren()
    {
        _proben.Clear();
        Speichern();
    }
}
